File: 프로그래머스/3/60059. 자물쇠와 열쇠/자물쇠와 열쇠.py
def turn_90degree(key):
    n = len(key)
    m = len(key[0])
    result = [[0] * n for _ in range(m)]
    
    for i in range(n):
        for j in range(m):
            result[j][n-i-1] = key[i][j]
    return result

def check(new_lock):
    lock_length = len(new_lock) // 3
    
    for i in range(lock_length, lock_length * 2):
        for j in range(lock_length, lock_length * 2):
            if new_lock[i][j] != 1:
                return False
            
    return True
            

def solution(key, lock):
    n = len(lock)
    m = len(key)
    new_lock = [[0] * (n*3) for _ in range(n*3)]
    
    for i in range(n):
        for j in range(n):
            new_lock[i+n][j+n] = lock[i][j]
            
    for _ in range(4):
        key = turn_90degree(key)
        
        for x in range(1, n*2):
            for y in range(1, n*2):

                for i in range(m):
                    for j in range(m):
                        new_lock[x+i][y+j] += key[i][j]

                if check(new_lock) == True:
                    return True

                for i in range(m):
                    for j in range(m):
                        new_lock[x+i][y+j] -= key[i][j]
            
    return False

File: 프로그래머스/3/60059. 자물쇠와 열쇠/자물쇠와 열쇠.java
import java.util.*;

class Solution {
    
    public boolean solution(int[][] key, int[][] lock) {
        int n = lock.length;
        int m = key.length;
        
        int[][] board = new int[n*3][n*3];
        
        for (int i = 0; i < n; i++){
            for (int j = 0; j < n; j++){
                board[i+n][j+n] = lock[i][j];
            }
        }
        
        for (int i = 1; i< n*2; i++){
            for(int j = 1; j < n*2; j++){
                for (int k = 0; k < 4; k++){
                    key = rotateKey(key);
                    
                    for (int x = 0; x < m; x++){
                        for (int y = 0; y < m; y++){
                            board[x + i][y + j] += key[x][y];
                        }
                    }
                    
                    if (check(board)){
                        return true;
                    }
                    
                    for (int x = 0; x < m; x++){
                        for (int y = 0; y < m; y++){
                            board[x + i][y + j] -= key[x][y];
                        }
                    }
                }
            }
        }
        return false;
    }
    
    static boolean check(int[][] board) {
        int lockLength = (board.length) / 3;
        for (int i = lockLength; i < lockLength * 2; i++){
            for (int j =lockLength; j < lockLength * 2; j++){
                if (board[i][j] != 1)
                    return false;
            }
        }
        return true;
    }
    
    static int[][] rotateKey(int[][] key){
        int n = key.length;
        int m = key[0].length;
        int[][] result = new int[m][n];
        
        for (int i = 0; i < n; i++){
            for (int j = 0; j < m; j++){
                result[j][n-1-i] = key[i][j];
            }
        }
        
        return result;
    }
}